src/transaction_builder.js

```diff
@@ -217,59 +280,119 @@ TransactionBuilder.prototype.__build = function(allowIncomplete) {
-    assert.notEqual(scriptType, 'nonstandard', 'RedeemScript not supported (nonstandard)')
+    // redeemScript only needed to initialize, but if provided again, enforce consistency
+    if (redeemScript) {
+      assert.deepEqual(input.redeemScript, redeemScript, 'Inconsistent redeemScript')
+    }
 
-    hash = this.tx.hashForSignature(index, redeemScript, hashType)
+    // if signatures already exist, enforce multisig scriptType
+    if (input.signatures.length > 0) {
+      assert.equal(input.scriptType, 'multisig', input.scriptType + ' doesn\'t support multiple signatures')
+    }
 
+  // initialize it
   } else {
-    prevOutScript = prevOutScript || privKey.pub.getAddress().toOutputScript()
-    prevOutType = prevOutType || 'pubkeyhash'
+    if (redeemScript) {
+      if (input.prevOutScript) {
+        assert.equal(input.prevOutType, 'scripthash', 'PrevOutScript must be P2SH')
 
-    assert.notEqual(prevOutType, 'scripthash', 'PrevOutScript is P2SH, missing redeemScript')
+        var scriptHash = input.prevOutScript.chunks[1]
+        assert.deepEqual(scriptHash, redeemScript.getHash(), 'RedeemScript does not match ' + scriptHash.toString('hex'))
 
-    scriptType = prevOutType
+      } else {
+        input.prevOutScript = scripts.scriptHashOutput(redeemScript.getHash())
+        input.prevOutType = 'scripthash'
+      }
 
-    hash = this.tx.hashForSignature(index, prevOutScript, hashType)
-  }
+      var scriptType = scripts.classifyOutput(redeemScript)
+
+      switch (scriptType) {
+        case 'multisig':
+          input.pubKeys = redeemScript.chunks.slice(1, -2).map(ECPubKey.fromBuffer)
+          break
+
+        case 'pubkeyhash':
+          var pkh1 = redeemScript.chunks[2]
+          var pkh2 = privKey.pub.getAddress().hash
+
+          assert.deepEqual(pkh1, pkh2, 'privateKey cannot sign for this input')
+          input.pubKeys = [privKey.pub]
+          break
 
-  this.prevOutScripts[index] = prevOutScript
-  this.prevOutTypes[index] = prevOutType
+        case 'pubkey':
+          input.pubKeys = redeemScript.chunks.slice(0, 1).map(ECPubKey.fromBuffer)
+          break
 
-  if (!(index in this.signatures)) {
-    this.signatures[index] = {
-      hashType: hashType,
-      pubKeys: [],
-      redeemScript: redeemScript,
-      scriptType: scriptType,
-      signatures: []
+        default:
+          assert(false, 'RedeemScript not supported (' + scriptType + ')')
+      }
+
+      input.redeemScript = redeemScript
+      input.scriptType = scriptType
+
+    } else {
+      assert.notEqual(input.prevOutType, 'scripthash', 'PrevOutScript is P2SH, missing redeemScript')
+
+      if (!input.scriptType) {
+        input.prevOutScript = privKey.pub.getAddress().toOutputScript()
+        input.prevOutType = 'pubkeyhash'
+        input.pubKeys = [privKey.pub]
+        input.scriptType = input.prevOutType
+      }
     }
+
+    input.hashType = hashType
+    input.initialized = true
```

Comment:
What, test that `initialized` is `true`?  I see it simply as an implementation artifact for now.  I guess I could test for it, but it could be removed at any time.